mediapipe/tasks/ios/genai/inference/sources/LlmInference+Session.swift

```diff
@@ -0,0 +1,291 @@
+      /// Disallow response generation if another response generation call initiated by any
+      /// `LlmInference` used to create the current session is already in progress.
+      ///
+      /// TODO: If simultaneous response generations on multiple sessions or the same session
+      /// are allowed to happen it leads to a crash. Investigate if this can be handled by C++.
+      try llmInference.shouldContinueWithResponseGeneration()
+
+      let tokens = try llmSessionRunner.predict()
+
+      llmInference.markResponseGenerationCompleted()
+
+      guard let humanReadableLlmResponse = Session.humanReadableString(llmResponses: tokens)
+      else {
+        throw GenAiInferenceError.invalidResponse
+      }
+
+      return humanReadableLlmResponse
+    }
+
+    /// Generates a response based on the previously added query chunks asynchronously. The
+    /// `progress` callback returns the partial responses from the LLM or any errors.
+    /// `completion` callback is invoked once the LLM is done generating responses.
+    /// Use `addQueryChunk(inputText:)` to add atleast one query chunk before calling this function.
+    /// Note: You cannot invoke simultaneous response generation calls on active sessions created
+    /// using the same `LlmInference`. You have to wait for the currently running response
+    /// generation call to complete before initiating another one.
+    ///
+    /// - Parameters:
+    ///   - progress: A callback invoked when a partial response is available from the LLM.
+    ///   - completion: A callback invoked when the LLM finishes response generation.
+    /// - Throws: An error if the LLM's response is invalid or if a response generation is
+    /// currently in progress on any session initialized from the `LlmInference` used to create
+    /// this session.
+    @objc public func generateResponseAsync(
+      progress: @escaping (_ partialResponse: String?, _ error: Error?) -> Void,
+      completion: @escaping (() -> Void)
+    ) throws {
+      /// Disallow response generation if another response generation call initiated by any
+      /// `LlmInference` used to create the current session is already in progress.
+      ///
+      /// TODO: If simultaneous response generations on multiple sessions or the same session
+      /// are allowed to happen it leads to a crash. Investigate if this can be handled by C++.
+      try llmInference.shouldContinueWithResponseGeneration()
+
+      /// Used to make a decision about whitespace stripping.
+      var receivedFirstToken = true
+
+      try llmSessionRunner.predictAsync(
+        progress: { partialResponseStrings, error in
+          guard let responseStrings = partialResponseStrings,
+            let humanReadableLlmResponse = Session.humanReadableString(
+              llmResponses: responseStrings, stripLeadingWhitespaces: receivedFirstToken)
+          else {
+            progress(nil, GenAiInferenceError.invalidResponse)
```

Comment:
We should also call `self?.llmInference.markResponseGenerationCompleted()` if it errors out?

Reply:
You mean if it errors out in the progress callback? I am assuming, even if one of the partial responses is invalid, the C++ session would still continue giving callbacks for the remaining responses? 
In that case wouldn't marking it completed, still cause the same issue? By design, irrespective of whether the response returned by C++ is invalid, the session runner will also call `completion()` after `progress()` if the response context marks the bool `done` as true.

LMK if this is not how C++ behaves.

Reply:
But thanks to your query, I spotted an issue in the control flow of sync `predict()` when an error happens and have fixed the same.

Reply:
Thanks for the explanation. I took a look at llmSessionRunner.predictAsync implementation, we don't need to call `self?.llmInference.markResponseGenerationCompleted()` here.